website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Table Scan: Doing The Right Thing With Structured Types
+author: Orri Erling
+authorURL: https://www.linkedin.com/in/orrierling/
+authorFBID: 100026224749124
+---
+
+In the previous article we saw what gains are possible when filtering early and in the right order. In this article we look at how we do this with nested and structured types.
+
+<!--truncate-->
+
+We use the 100G TPC-H dataset but now we group top level columns into structs or maps. 
+
+The tables are defined as follows:
+
+```sql
+CREATE TABLE hive.tpch.exportlineitem  (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    l_shipment row(
+        l_partkey BIGINT,
+        l_suppkey BIGINT,
+        l_extendedprice DOUBLE,
+        l_discount DOUBLE,
+        l_quantity DOUBLE,
+        l_shipdate DATE,
+        l_receiptdate DATE,
+        l_commitdate DATE,
+        l_comment VARCHAR(44)),
+    l_export row(
+        s_nation BIGINT,
+        c_nation BIGINT,
+        is_inside_eu INTEGER,
+        is_restricted INTEGER,
+        license row(
+            applydate DATE,
+            grantdate DATE,
+            filing_no BIGINT,
+            comment VARCHAR))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+The l_shipment struct has the non-null top-level columns we all have come to know and love. The l_export struct is null if the customer and supplier nations are the same and present otherwise. A fraction of the rows have an additional nested export license struct.
+
+```sql
+CREATE TABLE hive.tpch.lineitem_map (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    ints map(INTEGER, BIGINT),
+    strs map(VARCHAR(2), varchar(44))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+This table has a map of 12 integers in `ints` and 5 strings in `strs`. The key is the column ordinal number in the original lineitem table as an integer in `ints` and as a string in `strs`.
```

Comment:
- consistency: sometimes column names are spelled without back-ticks, sometimes with them.
- would it be possible to use more descriptive names for ints and strs?
- would you include SQL used to populate this table?

Reply:
Tried to make it all consistent, but left the names as is. SQL is in the linked branch.

It's a bit of a mess because it really is just ints and strs. The contents by index are all kinds of things and have no logical grouping.